### ISSUE TITLE
investigate mutable default parameters

### ISSUE THREAD

**Issue by sils1297** [coala-analyzer/coala#452]
we do have mutable default parameters on several occasions now. We should really investigate possible drawbacks of this "convenience".

**Comment by sils1297** [comment]
I'd love it if anyone could leave an opinion on this.

**Comment by sils1297** [comment]
@Makman2 @Udayan12167 @smfn3321 any opinions?

**Comment by Udayan12167** [comment]
By mutable default parameters you mean the defaults in SectionManager _fill_sections?

**Comment by sils1297** [comment]
@Udayan12167 SectionManager has no _fill_sections method. A short example:

```python
def foo(a=[]):
    a.append(5)
    return a
```

```
>>> foo()
[5]
>>> foo()
[5, 5]
>>> foo()
[5, 5, 5]
>>> foo()
[5, 5, 5, 5]
```

The object `[]`, is mutable and the 5 gets appended to it while it's "identity" remains and it is still the default value.

**Comment by Udayan12167** [comment]
My bad it _fill_settings anyways it is irrelevant. Can you give me an example in the coala code.

**Comment by sils1297** [comment]
```python
class CliParser(SectionParser):
    def __init__(self,
                 arg_parser=default_arg_parser,
                 key_value_delimiters=['=', ':'],
                 comment_seperators=[],
                 key_delimiters=[','],
                 section_override_delimiters=["."]):
```

**Comment by Udayan12167** [comment]
In this case it is assignment not appending right?

**Comment by sils1297** [comment]
@Udayan12167 the problem is that somebody might append something in some member function in the future without looking at the constructor and the implications. It makes things more complex and you have to look at every special case and some day we'll have a bug from it because someone doesn't think of it.

Thus I pledge to generically forbid mutable default values.

**Comment by Udayan12167** [comment]
https://pythonhosted.org/pysistence/

**Comment by sils1297** [comment]
@Udayan12167 I'm not sure if we want to use an extra library for this problem. I'd rather go for using `None` as param and then `self.member = param if param is not None else []` in the code

**Comment by sils1297** [comment]
http://stackoverflow.com/questions/1132941/least-astonishment-in-python-the-mutable-default-argument

**Comment by Makman2** [comment]
Basic defaults fixed in #773, if we find other mutable objects, we open a new issue for that^^
